{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Defining a function to create an artificial dataset with parameters, there will be one ideal/strong bin\n",
    "#Note: MAF (minor allele frequency) cutoff refers to the threshold separating rare variant features from common features\n",
    "def RVDS_One_Bin(number_of_instances, number_of_features, number_of_features_in_bin, \n",
    "                 rare_variant_MAF_cutoff, endpoint_cutoff_parameter, endpoint_variation_probability):\n",
    "    \n",
    "    #Creating an empty dataframe to use as a starting point for the eventual feature matrix\n",
    "    #Adding one to number of features to give space for the class column\n",
    "    df = pd.DataFrame(np.zeros((number_of_instances, number_of_features+1)))\n",
    "    \n",
    "    #Creating a list of features\n",
    "    feature_list = []\n",
    "    \n",
    "    #Creating a list of predictive features in the strong bin\n",
    "    predictive_features = []\n",
    "    for a in range (0, number_of_features_in_bin):\n",
    "        predictive_features.append(\"P_\" + str(a+1))\n",
    "    \n",
    "    for b in range (0, len(predictive_features)):\n",
    "        feature_list.append(predictive_features[b])\n",
    "            \n",
    "    #Creating a list of randomly created features\n",
    "    random_features = []\n",
    "    for c in range (0, number_of_features - number_of_features_in_bin):\n",
    "        random_features.append(\"R_\" + str(c+1))\n",
    "    \n",
    "    for d in range (0, len(random_features)):\n",
    "        feature_list.append(random_features[d])\n",
    "    \n",
    "    #Adding the features and the class/endpoint \n",
    "    features_and_class = feature_list.copy()\n",
    "    features_and_class.append('Class')\n",
    "    df.columns = features_and_class\n",
    "    \n",
    "    #Creating a list of numbers with the amount of numbers equal to the number of instances\n",
    "    #This will be used when assigning values to the values of features that are in the bin\n",
    "    instance_list = []\n",
    "    for number in range (0, number_of_instances):\n",
    "        instance_list.append(number)\n",
    "    \n",
    "    #ASSIGNING VALUES TO PREDICTIVE FEATURES\n",
    "    #Randomly assigning instances in each of the predictive features the value of 1 or 2\n",
    "    #Ensuring that the MAF (minor allele frequency) of each feature is a random value between 0 and the cutoff\n",
    "    #Multiplying by 2 because there are two alleles for each instance\n",
    "    for e in range (0, number_of_features_in_bin):\n",
    "        #Calculating the sum of instances with minor alleles\n",
    "        MA_sum = round((random.uniform(0, 2*rare_variant_MAF_cutoff))*number_of_instances)\n",
    "        #Between 0 and 50% of the minor allele sum will be from instances with value 2\n",
    "        number_of_MA2_instances = round(0.5*(random.uniform(0, MA_sum*0.5)))\n",
    "        #The remaining MA instances will have a value of 1\n",
    "        number_of_MA1_instances = MA_sum - 2*number_of_MA2_instances\n",
    "        MA1_instances = random.sample(instance_list, number_of_MA1_instances)\n",
    "        for f in MA1_instances:\n",
    "            df.at[f, predictive_features[e]] = 1\n",
    "        instances_wo_MA1 = list(set(instance_list) - set(MA1_instances))\n",
    "        MA2_instances = random.sample(instances_wo_MA1, number_of_MA2_instances)\n",
    "        for f in MA2_instances:\n",
    "            df.at[f, predictive_features[e]] = 2\n",
    "    \n",
    "    #ASSIGNING ENDPOINT (CLASS) VALUES \n",
    "    #Creating a list of bin values for the sum of values across predictive features in the bin\n",
    "    bin_values = []\n",
    "    for g in range (0, number_of_instances):\n",
    "        sum_of_values = 0\n",
    "        for h in predictive_features:\n",
    "            sum_of_values += df.iloc[g][h]\n",
    "        bin_values.append(sum_of_values)\n",
    "    \n",
    "    #User input for the cutoff between 0s and 1s is either mean or median. \n",
    "    if endpoint_cutoff_parameter == 'mean':\n",
    "        #Finding the mean to make a cutoff for 0s and 1s in the class colum\n",
    "        endpoint_cutoff = statistics.mean(bin_values)\n",
    "    \n",
    "        #If the sum of feature values in an instance is greater than or equal to the mean, then the endpoint will be 1\n",
    "        for i in range (0, number_of_instances):\n",
    "            if bin_values[i] > endpoint_cutoff:\n",
    "                df.at[i, 'Class'] = 1\n",
    "        \n",
    "            elif bin_values[i] == endpoint_cutoff:\n",
    "                df.at[i, 'Class'] = 1\n",
    "            \n",
    "            elif bin_values[1] < endpoint_cutoff:\n",
    "                df.at[i, 'Class'] = 0\n",
    "    \n",
    "    elif endpoint_cutoff_parameter == 'median':\n",
    "        #Finding the median to make a cutoff for 0s and 1s in the class colum\n",
    "        endpoint_cutoff = statistics.median(bin_values)\n",
    "    \n",
    "        #If the sum of feature values in an instance is greater than or equal to the mean, then the endpoint will be 1\n",
    "        for i in range (0, number_of_instances):\n",
    "            if bin_values[i] > endpoint_cutoff:\n",
    "                df.at[i, 'Class'] = 1\n",
    "        \n",
    "            elif bin_values[i] == endpoint_cutoff:\n",
    "                df.at[i, 'Class'] = 1\n",
    "            \n",
    "            elif bin_values[1] < endpoint_cutoff:\n",
    "                df.at[i, 'Class'] = 0\n",
    "        \n",
    "    #Applying the \"noise\" parameter to introduce endpoint variability\n",
    "    for instance in range (0, number_of_instances):\n",
    "        if endpoint_variation_probability > random.uniform(0,1):\n",
    "            if df.loc[instance]['Class'] == 0:\n",
    "                df.at[instance, 'Class'] = 1\n",
    "            \n",
    "            elif df.loc[instance]['Class'] == 1:\n",
    "                df.at[instance,'Class'] = 0\n",
    "    \n",
    "    #ASSIGNING VALUES TO RANDOM FEATURES\n",
    "    #Randomly assigning instances in each of the random features the value of 1\n",
    "    #Ensuring that the MAF of each feature is a random value between 0 and the cutoff (probably 0.05)\n",
    "    #Multiplying by 2 because there are two alleles for each instance\n",
    "    for e in range (0, len(random_features)):\n",
    "        #Calculating the sum of instances with minor alleles\n",
    "        MA_sum = round((random.uniform(0, 2*rare_variant_MAF_cutoff))*number_of_instances)\n",
    "        #Between 0 and 50% of the minor allele sum will be from instances with value 2\n",
    "        number_of_MA2_instances = round(0.5*(random.uniform(0, MA_sum*0.5)))\n",
    "        #The remaining MA instances will have a value of 1\n",
    "        number_of_MA1_instances = MA_sum - 2*number_of_MA2_instances\n",
    "        MA1_instances = random.sample(instance_list, number_of_MA1_instances)\n",
    "        for f in MA1_instances:\n",
    "            df.at[f, random_features[e]] = 1\n",
    "        instances_wo_MA1 = list(set(instance_list) - set(MA1_instances))\n",
    "        MA2_instances = random.sample(instances_wo_MA1, number_of_MA2_instances)\n",
    "        for f in MA2_instances:\n",
    "            df.at[f, random_features[e]] = 2\n",
    "    \n",
    "    return df, endpoint_cutoff\n",
    "\n",
    "#Example of calling RVDS_One_Bin Function:\n",
    "# df, endpoint_cutoff = RVDS_One_Bin(1000, 50, 10, 0.05, 'mean', 0):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Defining a function to create an artificial dataset with parameters\n",
    "#There will be an epistatic relationship between a bin and a common feature\n",
    "#Note: MAF (minor allele frequency) cutoff refers to the threshold separating rare variant features from common features\n",
    "#Common feature genotype frequencies list should be of the form [0.25, 0.5, 0.25] (numbers should add up to 1)\n",
    "#List of predictive MLGs (multi-locus genotypes) should contain any of the 9 possibilites and be of the form below:\n",
    "#[AABB, AABb, AAbb, AaBB, AaBb, Aabb, aaBB, aaBb, aabb]\n",
    "#Genotype Cutoff Metric can be mean or median\n",
    "def RVDS_Bin_Epistatic_Interaction_with_Common_Feature(number_of_instances, number_of_rare_features, \n",
    "                                                       number_of_features_in_bin, \n",
    "                                                       rare_variant_MAF_cutoff, common_feature_genotype_frequencies_list, \n",
    "                                                       genotype_cutoff_metric, \n",
    "                                                       endpoint_variation_probability,\n",
    "                                                       list_of_MLGs_predicting_disease, print_summary):\n",
    "    \n",
    "    #Creating an empty dataframe to use as a starting point for the eventual feature matrix\n",
    "    #Adding two to number of rare features to give space for the class column and give space for the common feature\n",
    "    df = pd.DataFrame(np.zeros((number_of_instances, number_of_rare_features+2)))\n",
    "    \n",
    "    #Creating a list of features\n",
    "    feature_list = []\n",
    "    \n",
    "    #Creating a list of features in the bin that interacts epistatically with the common feature\n",
    "    predictive_features = []\n",
    "    for a in range (0, number_of_features_in_bin):\n",
    "        predictive_features.append(\"P_\" + str(a+1))\n",
    "    \n",
    "    for b in range (0, len(predictive_features)):\n",
    "        feature_list.append(predictive_features[b])\n",
    "            \n",
    "    #Creating a list of randomly created features\n",
    "    random_features = []\n",
    "    for c in range (0, number_of_rare_features - number_of_features_in_bin):\n",
    "        random_features.append(\"R_\" + str(c+1))\n",
    "    \n",
    "    for d in range (0, len(random_features)):\n",
    "        feature_list.append(random_features[d])\n",
    "    \n",
    "    #Adding the common feature to the feature list\n",
    "    feature_list.append('Common_Feature')\n",
    "    \n",
    "    #Adding the features and the class/endpoint \n",
    "    features_and_class = feature_list\n",
    "    features_and_class.append('Class')\n",
    "    df.columns = features_and_class\n",
    "    \n",
    "    #Creating a list of numbers with the amount of numbers equal to the number of instances\n",
    "    #This will be used when assigning values to the values of features that are in the bin\n",
    "    instance_list = []\n",
    "    for number in range (0, number_of_instances):\n",
    "        instance_list.append(number)\n",
    "    \n",
    "    #RANDOMLY ASSIGNING VALUES TO THE FEATURES IN THE BIN\n",
    "    #Randomly assigning instances in each of the predictive features the value of 1 or 2\n",
    "    #Ensuring that the MAF (minor allele frequency) of each feature is a random value between 0 and the cutoff\n",
    "    for e in range (0, number_of_features_in_bin):\n",
    "        #Calculating the sum of instances with minor alleles\n",
    "        MA_sum = round((random.uniform(0, 2*rare_variant_MAF_cutoff))*number_of_instances)\n",
    "        #Between 0 and 50% of the minor allele sum will be from instances with value 2\n",
    "        number_of_MA2_instances = round(0.5*(random.uniform(0, MA_sum*0.5)))\n",
    "        #The remaining MA instances will have a value of 1\n",
    "        number_of_MA1_instances = MA_sum - 2*number_of_MA2_instances\n",
    "        MA1_instances = random.sample(instance_list, number_of_MA1_instances)\n",
    "        for f in MA1_instances:\n",
    "            df.at[f, predictive_features[e]] = 1\n",
    "        instances_wo_MA1 = list(set(instance_list) - set(MA1_instances))\n",
    "        MA2_instances = random.sample(instances_wo_MA1, number_of_MA2_instances)\n",
    "        for f in MA2_instances:\n",
    "            df.at[f, predictive_features[e]] = 2\n",
    "    \n",
    "    #DETERMING GENOTYPE CUTOFF FOR THE BIN (AA vs. Aa vs. aa) \n",
    "    #Creating a list of bin values for the sum of values across predictive features in the bin\n",
    "    bin_values = []\n",
    "    for g in range (0, number_of_instances):\n",
    "        sum_of_values = 0\n",
    "        for h in predictive_features:\n",
    "            sum_of_values += df.iloc[g][h]\n",
    "        bin_values.append(sum_of_values)\n",
    "    \n",
    "    #Bin values that are 0 will be AA\n",
    "    \n",
    "    nonzero_bin_values = []\n",
    "    #Creating a list of nonzero bin values, the mean of median of this will determine the cutoff between Aa and aa\n",
    "    for i in range (0, len(bin_values)):\n",
    "        if bin_values[i] != 0:\n",
    "            nonzero_bin_values.append(bin_values[i])\n",
    "    \n",
    "    if genotype_cutoff_metric == 'mean':\n",
    "        Aa_aa_cutoff = statistics.mean(nonzero_bin_values)\n",
    "        \n",
    "    elif genotype_cutoff_metric == 'median':\n",
    "        Aa_aa_cutoff = statistics.median(nonzero_bin_values)\n",
    "    \n",
    "    #Creating a list for each of the bin genotypes\n",
    "    bin_AA_genotype_list = []\n",
    "    bin_Aa_genotype_list = []\n",
    "    bin_aa_genotype_list = []\n",
    "    \n",
    "    for g in range (0, number_of_instances):\n",
    "        if bin_values[g] == 0:\n",
    "            bin_AA_genotype_list.append(g)\n",
    "        \n",
    "        elif bin_values[g] > 0  and bin_values[g] < Aa_aa_cutoff:\n",
    "            bin_Aa_genotype_list.append(g)\n",
    "            \n",
    "        elif bin_values[g] == Aa_aa_cutoff or bin_values[g] > Aa_aa_cutoff:\n",
    "            bin_aa_genotype_list.append(g)\n",
    "    \n",
    "    #ASSIGNING VALUES FOR THE COMMON FEATURE \n",
    "    #Based on the given allele frequencies the user inputs, randomly choosing instances for each genotype (BB, Bb, bb)\n",
    "    #of the common feature\n",
    "    \n",
    "    instances_left = instance_list.copy()\n",
    "    common_feature_BB_genotype_list = random.sample(instance_list, (round(number_of_instances * float(common_feature_genotype_frequencies_list[0]))))\n",
    "    for instance in common_feature_BB_genotype_list:\n",
    "        df.at[instance, 'Common_Feature'] = 0\n",
    "        instances_left.remove(instance)\n",
    "    \n",
    "    common_feature_Bb_genotype_list = random.sample(instances_left, (round(number_of_instances * float(common_feature_genotype_frequencies_list[1]))))\n",
    "    for instance in common_feature_Bb_genotype_list:\n",
    "        df.at[instance, 'Common_Feature'] = 1\n",
    "        instances_left.remove(instance)\n",
    "    \n",
    "    common_feature_bb_genotype_list = random.sample(instances_left, (round(number_of_instances * float(common_feature_genotype_frequencies_list[2]))))\n",
    "    for instance in common_feature_bb_genotype_list:\n",
    "        df.at[instance, 'Common_Feature'] = 2\n",
    "    \n",
    "    #ASSIGNING CLASS/ENDPOINT VALUES \n",
    "    \n",
    "    #First assigning a value of 0 for all the classes (this will be overwrote later)\n",
    "    for instance in instance_list:\n",
    "        df.at[instance, 'Class'] = 0\n",
    "        \n",
    "    #Assigning a value of 1 for each instance that matches the MLG (multi-locus genotype) that the user specifies should\n",
    "    #result in a diseased state\n",
    "    disease_instances = []\n",
    "    for instance in instance_list:\n",
    "        if 'AABB' in list_of_MLGs_predicting_disease:\n",
    "            if instance in bin_AA_genotype_list and instance in common_feature_BB_genotype_list:\n",
    "                df.at[instance, 'Class'] = 1\n",
    "                disease_instances.append(instance)\n",
    "        \n",
    "        if 'AABb' in list_of_MLGs_predicting_disease:\n",
    "            if instance in bin_AA_genotype_list and instance in common_feature_Bb_genotype_list:\n",
    "                df.at[instance, 'Class'] = 1\n",
    "                disease_instances.append(instance)\n",
    "                \n",
    "        if 'AAbb' in list_of_MLGs_predicting_disease:\n",
    "            if instance in bin_AA_genotype_list and instance in common_feature_bb_genotype_list:\n",
    "                df.at[instance, 'Class'] = 1\n",
    "                disease_instances.append(instance)\n",
    "        \n",
    "        if 'AaBB' in list_of_MLGs_predicting_disease:\n",
    "            if instance in bin_Aa_genotype_list and instance in common_feature_BB_genotype_list:\n",
    "                df.at[instance, 'Class'] = 1\n",
    "                disease_instances.append(instance)\n",
    "                \n",
    "        if 'AaBb' in list_of_MLGs_predicting_disease:\n",
    "            if instance in bin_Aa_genotype_list and instance in common_feature_Bb_genotype_list:\n",
    "                df.at[instance, 'Class'] = 1\n",
    "                disease_instances.append(instance)\n",
    "        \n",
    "        if 'Aabb' in list_of_MLGs_predicting_disease:\n",
    "            if instance in bin_Aa_genotype_list and instance in common_feature_bb_genotype_list:\n",
    "                df.at[instance, 'Class'] = 1\n",
    "                disease_instances.append(instance)\n",
    "\n",
    "        if 'aaBB' in list_of_MLGs_predicting_disease:\n",
    "            if instance in bin_aa_genotype_list and instance in common_feature_BB_genotype_list:\n",
    "                df.at[instance, 'Class'] = 1\n",
    "                disease_instances.append(instance)\n",
    "        \n",
    "        if 'aaBb' in list_of_MLGs_predicting_disease:\n",
    "            if instance in bin_aa_genotype_list and instance in common_feature_Bb_genotype_list:\n",
    "                df.at[instance, 'Class'] = 1\n",
    "                disease_instances.append(instance)\n",
    "        \n",
    "        if 'aabb' in list_of_MLGs_predicting_disease:\n",
    "            if instance in bin_aa_genotype_list and instance in common_feature_bb_genotype_list:\n",
    "                df.at[instance, 'Class'] = 1\n",
    "                disease_instances.append(instance)\n",
    "        \n",
    "    #Applying the \"noise\" parameter to introduce endpoint variability\n",
    "    for instance in range (0, number_of_instances):\n",
    "        if endpoint_variation_probability > random.uniform(0,1):\n",
    "            if df.loc[instance]['Class'] == 0:\n",
    "                df.at[instance, 'Class'] = 1\n",
    "            \n",
    "            elif df.loc[instance]['Class'] == 1:\n",
    "                df.at[instance,'Class'] = 0\n",
    "    \n",
    "    #ASSIGNING VALUES TO RANDOM FEATURES\n",
    "    #Randomly assigning instances in each of the random features the value of 1\n",
    "    #Ensuring that the MAF of each feature is a random value between 0 and the cutoff (probably 0.05)\n",
    "    for e in range (0, len(random_features)):\n",
    "        #Calculating the sum of instances with minor alleles\n",
    "        MA_sum = round((random.uniform(0, 2*rare_variant_MAF_cutoff))*number_of_instances)\n",
    "        #Between 0 and 50% of the minor allele sum will be from instances with value 2\n",
    "        number_of_MA2_instances = round(0.5*(random.uniform(0, MA_sum*0.5)))\n",
    "        #The remaining MA instances will have a value of 1\n",
    "        number_of_MA1_instances = MA_sum - 2*number_of_MA2_instances\n",
    "        MA1_instances = random.sample(instance_list, number_of_MA1_instances)\n",
    "        for f in MA1_instances:\n",
    "            df.at[f, random_features[e]] = 1\n",
    "        instances_wo_MA1 = list(set(instance_list) - set(MA1_instances))\n",
    "        MA2_instances = random.sample(instances_wo_MA1, number_of_MA2_instances)\n",
    "        for f in MA2_instances:\n",
    "            df.at[f, random_features[e]] = 2\n",
    "    \n",
    "    if print_summary == True:\n",
    "        #PRINTING INFORMATION ABOUT THE DATASET THAT CAN BE INPUTTED IN A PENETRANCE TABLE\n",
    "        print(\"Probability of Disease (Class = 1) for Multi-Locus Genotypes:\")\n",
    "        MLG_list = ['AABB', 'AABb', 'AAbb', 'AaBB', 'AaBb', 'Aabb', 'aaBB', 'aaBb', 'aabb']\n",
    "        MLG_penetrance_dict = {}\n",
    "        for MLG in MLG_list:\n",
    "            if MLG in list_of_MLGs_predicting_disease:\n",
    "                print(MLG + ': 1')\n",
    "                MLG_penetrance_dict[MLG] = 1\n",
    "            elif MLG not in list_of_MLGs_predicting_disease:\n",
    "                print(MLG + ': 0')\n",
    "                MLG_penetrance_dict[MLG] = 0\n",
    "        print(\"---\")\n",
    "        print('Marginal Penetrance for Genotypes:')\n",
    "        AA_penetrance = ((len(common_feature_BB_genotype_list)*MLG_penetrance_dict['AABB']/number_of_instances) + \n",
    "                          (len(common_feature_Bb_genotype_list)*MLG_penetrance_dict['AABb']/number_of_instances) + \n",
    "                          (len(common_feature_bb_genotype_list)*MLG_penetrance_dict['AAbb']/number_of_instances))\n",
    "        print(\"AA: \" + str(AA_penetrance))\n",
    "    \n",
    "        Aa_penetrance = ((len(common_feature_BB_genotype_list)*MLG_penetrance_dict['AaBB']/number_of_instances) + \n",
    "                          (len(common_feature_Bb_genotype_list)*MLG_penetrance_dict['AaBb']/number_of_instances) + \n",
    "                          (len(common_feature_bb_genotype_list)*MLG_penetrance_dict['Aabb']/number_of_instances))\n",
    "        print(\"Aa: \" + str(Aa_penetrance))\n",
    "          \n",
    "        aa_penetrance = ((len(common_feature_BB_genotype_list)*MLG_penetrance_dict['aaBB']/number_of_instances) + \n",
    "                          (len(common_feature_Bb_genotype_list)*MLG_penetrance_dict['aaBb']/number_of_instances) + \n",
    "                          (len(common_feature_bb_genotype_list)*MLG_penetrance_dict['aabb']/number_of_instances))\n",
    "        print(\"aa: \" + str(aa_penetrance))\n",
    "    \n",
    "        BB_penetrance = ((len(bin_AA_genotype_list)*MLG_penetrance_dict['AABB']/number_of_instances) + \n",
    "                          (len(bin_Aa_genotype_list)*MLG_penetrance_dict['AaBB']/number_of_instances) + \n",
    "                          (len(bin_aa_genotype_list)*MLG_penetrance_dict['aaBB']/number_of_instances))\n",
    "        print(\"BB: \" + str(BB_penetrance))\n",
    "    \n",
    "        Bb_penetrance = ((len(bin_AA_genotype_list)*MLG_penetrance_dict['AABb']/number_of_instances) + \n",
    "                          (len(bin_Aa_genotype_list)*MLG_penetrance_dict['AaBb']/number_of_instances) + \n",
    "                          (len(bin_aa_genotype_list)*MLG_penetrance_dict['aaBb']/number_of_instances))\n",
    "        print(\"Bb: \" + str(Bb_penetrance))\n",
    "    \n",
    "        bb_penetrance = ((len(bin_AA_genotype_list)*MLG_penetrance_dict['AAbb']/number_of_instances) + \n",
    "                          (len(bin_Aa_genotype_list)*MLG_penetrance_dict['Aabb']/number_of_instances) + \n",
    "                          (len(bin_aa_genotype_list)*MLG_penetrance_dict['aabb']/number_of_instances))\n",
    "        print(\"bb: \" + str(bb_penetrance))\n",
    "        print('---')\n",
    "        print(\"Population Prevalence of Disease (K): \" + str(len(disease_instances)/number_of_instances))\n",
    "    \n",
    "    return df, Aa_aa_cutoff\n",
    "\n",
    "#Example of calling RVDS_Bin_Epistatic_Interaction_with_Common_Feature function\n",
    "# df, Aa_aa_cutoff = RVDS_Bin_Epistatic_Interaction_with_Common_Feature(1000, 50, 10, 0.05, [0.25, 0.5, 0.25], \n",
    "#                                                                       'mean', 0, ['AABB', 'AaBb', 'aabb'], True):"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
